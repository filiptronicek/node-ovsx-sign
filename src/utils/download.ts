import fetch from "node-fetch";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

export const download = async (url: string, options: { filename?: string; foldername?: string }) => {
    const tmpDir = await fs.promises.mkdtemp(path.join(os.tmpdir(), "ovsx-"));
    const destinationFolder = options.foldername || tmpDir;

    const filename = options.filename || url.split("/").pop() || "downloaded-file";
    const destination = path.join(destinationFolder, filename);
    console.debug(`Downloading file from ${url} to ${destination}`);

    const response = await fetch(url);
    const fileStream = fs.createWriteStream(destination);
    await new Promise((resolve, reject) => {
        response.body.pipe(fileStream);
        response.body.on("error", reject);
        fileStream.on("finish", resolve);
    });
    return destination;
};
