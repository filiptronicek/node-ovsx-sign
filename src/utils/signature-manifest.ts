import { readZip } from "@vscode/vsce/out/zip";
import { createHash } from "node:crypto";
import * as fs from "node:fs/promises";

type VsixSignatureManifest = {
    package: {
        size: number;
        digests: {
            sha256: string;
        };
    };
    entries: {
        [key: string]: {
            size: number;
            digests: {
                sha256: string;
            };
        };
    };
};

/**
 * @returns a sha256 hash of the file
 */
const fileSha256 = async (file: Buffer): Promise<string> => {
    return new Promise((resolve, reject) => {
        try {
            const hash = createHash("sha256");
            hash.update(file);
            const digest = hash.digest("hex");
            resolve(digest);
        } catch (error) {
            reject(error);
        }
    });
};

const toBase64 = (string: string): string => Buffer.from(string).toString("base64");

export const generateSignatureManifest = async (vsixFilePath: string): Promise<VsixSignatureManifest> => {
    const extensionPackage = await fs.readFile(vsixFilePath);
    const extensionPackageContentsDetails: Map<string, Buffer> = await readZip(vsixFilePath, (entry) => true);

    const entries: VsixSignatureManifest["entries"] = {};
    for (const [entryPath, entryContent] of extensionPackageContentsDetails.entries()) {
        const entrySize = entryContent.length;
        const entrySha256 = await fileSha256(entryContent);
        entries[toBase64(entryPath)] = {
            size: entrySize,
            digests: {
                sha256: entrySha256,
            },
        };
    }

    return {
        package: {
            size: extensionPackage.length,
            digests: {
                sha256: await fileSha256(extensionPackage),
            },
        },
        entries,
    };
};
