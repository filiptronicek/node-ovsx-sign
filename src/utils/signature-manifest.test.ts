import { download } from "./download";
import { unlink } from "fs/promises";
import { generateSignatureManifest } from "./signature-manifest";

jest.setTimeout(20_000);

describe("signatureManifestTest", () => {
    test("be able to generate a signature manifest file", async () => {
        const extensionDestination = await download(
            "https://open-vsx.org/api/jeanp413/open-remote-ssh/0.0.33/file/jeanp413.open-remote-ssh-0.0.33.vsix",
            {},
        );
        expect(await generateSignatureManifest(extensionDestination)).toMatchSnapshot();

        // Clean up
        await unlink(extensionDestination);
        console.log(`Deleted ${extensionDestination}`);
    });
});
