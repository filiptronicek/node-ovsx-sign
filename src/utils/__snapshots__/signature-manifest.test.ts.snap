// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signatureManifestTest be able to generate a signature manifest file 1`] = `
{
  "entries": {
    "W2NvbnRlbnRfdHlwZXNdLnhtbA==": {
      "digests": {
        "sha256": "9320e1ab24aa3ce67af21fb9c59035a59a860752570ff02407c664a5b675027a",
      },
      "size": 462,
    },
    "ZXh0ZW5zaW9uL291dC9leHRlbnNpb24uanM=": {
      "digests": {
        "sha256": "1af5c4728d2422b0da6e3e79303c1d31bbcf22655233669a37f697ea739cac30",
      },
      "size": 535393,
    },
    "ZXh0ZW5zaW9uL2NoYW5nZWxvZy5tZA==": {
      "digests": {
        "sha256": "353ea85b16931f8fe064f3c89088cf24a0e57ec2db37a93506545d40f8525347",
      },
      "size": 277,
    },
    "ZXh0ZW5zaW9uL2xpY2Vuc2UudHh0": {
      "digests": {
        "sha256": "a5e6ddb3af783a02eacc848cbe309cc0940b4859ac653e9b1a2e21a4fea22552",
      },
      "size": 1056,
    },
    "ZXh0ZW5zaW9uL3BhY2thZ2UuanNvbg==": {
      "digests": {
        "sha256": "8244c0d61d31395969b111043ed842edda56f621a3419d49e3e097dba676428e",
      },
      "size": 10829,
    },
    "ZXh0ZW5zaW9uL3JlYWRtZS5tZA==": {
      "digests": {
        "sha256": "639a6272fcd5d082485afa3608dabc23eab7a0040c764de92c668d7666ca5f20",
      },
      "size": 1014,
    },
    "ZXh0ZW5zaW9uL3Jlc291cmNlcy9pY29uLnBuZw==": {
      "digests": {
        "sha256": "7dda10c951ece86295688e6126904a8f177b0b5068979b02db7ae64f71231305",
      },
      "size": 17422,
    },
    "ZXh0ZW5zaW9uLnZzaXhtYW5pZmVzdA==": {
      "digests": {
        "sha256": "04723002cd34ea52ed00181dec4278a88d9eed2eb219b1735d8be52fffa7b7c5",
      },
      "size": 2681,
    },
  },
  "package": {
    "digests": {
      "sha256": "3e45182da389bf4748decec32587315d89732e7cfa371a4bbff105616afc6e86",
    },
    "size": 173087,
  },
}
`;
